template/template.go

```diff
@@ -296,3 +303,218 @@ func (e *executer) resolveOptions(options []ast.Option) (map[string]any, error)
 
 	return m, nil
 }
+
+func (e *executer) resolveMatcher(m ast.Matcher) error {
+	res, err := e.resolveSelector(m)
+	if err != nil {
+		return fmt.Errorf("resolve selector: %w", err)
+	}
+
+	pref := e.resolvePreferences(m, res)
+
+	filteredVariants := e.filterVariants(m, pref)
+
+	sortable := e.sortVariants(filteredVariants, pref)
+
+	err = e.selectBestVariant(sortable)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *executer) resolveSelector(matcher ast.Matcher) ([]any, error) {
+	res := make([]any, 0, len(matcher.MatchStatements))
+
+	for _, selectExpr := range matcher.MatchStatements {
+		var function ast.Function
+
+		switch annotation := selectExpr.Annotation.(type) {
+		case nil:
+			return nil, ErrMissingSelectorAnnotation
+		case ast.ReservedAnnotation, ast.PrivateUseAnnotation:
+			return nil, ErrUnsupportedExpression
+		case ast.Function:
+			function = annotation
+		}
+
+		registryF, ok := e.template.funcRegistry[function.Identifier.Name]
+		if !ok {
+			return nil, fmt.Errorf("%w '%s'", ErrUnknownFunction, function.Identifier.Name)
+		}
+
+		opts, err := e.resolveOptions(function.Options)
+		if err != nil {
+			return nil, fmt.Errorf("resolve options: %w", err)
+		}
+
+		input, err := e.resolveValue(selectExpr.Operand)
+		if err != nil {
+			return nil, fmt.Errorf("resolve value: %w", err)
+		}
+
+		rslt, err := registryF.Match(input, opts)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %s", ErrSelection, err.Error())
+		}
+
+		res = append(res, rslt)
+	}
+
+	return res, nil
+}
+
+func (e *executer) resolvePreferences(m ast.Matcher, res []any) [][]string {
+	// Step 2: Resolve Preferences
+	pref := make([][]string, 0, len(res))
+
+	for i := range res {
+		var keys []string
+
+		for _, variant := range m.Variants {
+			for _, vKey := range variant.Keys {
+				switch key := vKey.(type) {
+				case ast.CatchAllKey:
+					continue
+				case ast.NameLiteral:
+					keys = append(keys, string(key))
+				case ast.QuotedLiteral:
+					keys = append(keys, string(key))
+				case ast.NumberLiteral:
+					keys = append(keys, fmt.Sprint(key))
+				}
+			}
+		}
+
+		rv := res[i]
+
+		matches := matchSelectorKeys(rv, keys)
+		pref = append(pref, matches)
+	}
+
+	return pref
+}
+
+func (e *executer) filterVariants(m ast.Matcher, pref [][]string) []ast.Variant {
+	// Step 3: Filter Variants
+	var filteredVariants []ast.Variant
+
+	for _, variant := range m.Variants {
+		matchesAllSelectors := true
+
+		for i, keyOrder := range pref {
+			key := variant.Keys[i]
+
+			var ks string
+
+			switch key := key.(type) {
+			case ast.CatchAllKey:
+				continue
+			case ast.NameLiteral:
+				ks = string(key)
+			case ast.QuotedLiteral:
+				ks = string(key)
```

Comment:
```suggestion
			case ast.NameLiteral, ast.QuotedLiteral:
				ks = string(key)
```

Reply:
is it allowed ?
If i remember correctly, type switch, with multiple types, then type is any, e.g. in this case key is any.
And you cannot cast any to string

Reply:
@VladislavsPerkanuks I resolved it by using a key.String(). I know that there is a behavior difference between string(key) and key.String(), but for now tests pass. Any objections?

Reply:
I will close eyes 🙈

Reply:
The underlying type of NameLiteral and QuotedLiteral is string. Always. Right?

It should not be possible to have a different type.

@VladislavsPerkanuks, under what circumstances can it be a different type?

Reply:
they are both string yes, but this is not possible anyways
```go
case ast.NameLiteral, ast.QuotedLiteral:
  ks = string(key) // cannot convert key (variable of type parse.VariantKey) to type string
```

in type switch, when there is multiple types in one case, it fallbacks to the type its being switched to, in this case VariantKey

Reply:
which is interface